src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs

```diff
@@ -0,0 +1,242 @@
+    {
+        CacheKey key = new CacheKey(settings, flags, certificate, intermediates, alpnProtocols, allowedCipherSuites);
+
+        MsQuicConfigurationSafeHandle? handle;
+
+        if (s_configurationCache.TryGetValue(key, out handle) && handle.TryAddRentCount())
+        {
+            if (NetEventSource.Log.IsEnabled())
+            {
+                NetEventSource.Info(null, $"Found cached MsQuicConfiguration: {handle}.");
+            }
+            return handle;
+        }
+
+        // if we get here, the handle is either not in the cache, or we lost the race between
+        // TryAddRentCount on this thread and MarkForDispose on another thread doing cache cleanup.
+        // In either case, we need to create a new handle.
+
+        if (NetEventSource.Log.IsEnabled())
+        {
+            NetEventSource.Info(null, $"MsQuicConfiguration not found in cache, creating new.");
+        }
+
+        handle = CreateInternal(settings, flags, certificate, intermediates, alpnProtocols, allowedCipherSuites);
+        handle.TryAddRentCount(); // we are the first renter
+
+        MsQuicConfigurationSafeHandle cached;
+        do
+        {
+            cached = s_configurationCache.GetOrAdd(key, handle);
+        }
+        // If we get the same handle back, we successfully added it to the cache and we are done.
+        // If we get a different handle back, we need to increase the rent count.
+        // If we fail to add the rent count, then the existing/cached handle is in process of
+        // being removed from the cache and we can try again, eventually either succeeding to add our
+        // new handle or getting a fresh handle inserted by another thread meanwhile.
+        while (cached != handle && !cached.TryAddRentCount());
+
+        if (cached != handle)
+        {
+            // we lost a race with another thread to insert new handle into the cache
+            if (NetEventSource.Log.IsEnabled())
+            {
+                NetEventSource.Info(null, $"Discarding MsQuicConfiguration {handle} (preferring cached {cached}).");
+            }
+
+            // First dispose decrements the rent count we added before attempting the cache insertion
+            // and second closes the handle
+            handle.Dispose();
+            handle.Dispose();
+            Debug.Assert(handle.IsClosed);
+
+            return cached;
+        }
+
+        // we added a new handle, check if we need to cleanup
+        if (s_configurationCache.Count % CheckExpiredModulo == 0)
+        {
+            // let only one thread perform cleanup at a time
+            lock (s_configurationCache)
+            {
+                if (s_configurationCache.Count % CheckExpiredModulo == 0)
```

Comment:
Do we need the second check here? If entries were being added quickly, this could mean we'll skip doing the cleanup more often.

Reply:
Right, as-is, this may skip cleanup, I think we just need to change the second condition to skip the cleanup only if the cache size decreased while we were taking the lock (meaning the cache was just cleaned)